White threshold for scanned PDFs

kerivin opened the pull request:
This PR is currently open as #15395

I left a comment in issue [#13294](https://github.com/koreader/koreader/issues/13294#issuecomment-2994186500) about brightness/contrast of scanned PDFs, but then I saw the amount of open issues and decided to experiment myself.

I added 'white threshold' option for MuPDF (indirectly) and k2pdfopt rendering that makes scanned PDFs way more readable on e-ink, since they often have gray background that becomes even darker with higher contrast, and with low contrast gray text on gray background is quite inconvenient to read, and I wish every reader had a feature like this.

See the [`koreader-base` pull request](https://github.com/koreader/koreader-base/pull/2115) and [`libk2pdfopt` pull request](https://github.com/koreader/libk2pdfopt/pull/62). I guess the optimal order of PRs would be `libk2pdfopt` -> update `libk2pdfopt` commit hash in `koreader-base` -> `koreader-base` -> update `koreader-base` submodule in `koreader` -> `koreader`.

I added the option for both k2pdfopt (native support) and MuPDF (custom implementation) because the feature seems to be useful regardless of whether 'dewatermark' is on or off.

via custom implementation for MuPDF:

<img src="https://github.com/user-attachments/assets/b150de06-fdd1-4310-a835-6cb7dc660625" width="300" />

<img src="https://github.com/user-attachments/assets/1b62f8d8-a7ce-45d3-b1ff-55d36277062f" width="300" />

<img src="https://github.com/user-attachments/assets/53deb03d-634f-4610-907e-ad4e8e9df1a3" width="600" />

via native k2pdfopt:

<img src="https://github.com/user-attachments/assets/57973405-e9bd-48f3-8dae-d923ea14c419" width="300" />

<img src="https://github.com/user-attachments/assets/02273398-88d4-4c9e-87f1-23988e1eaca6" width="300" />

<img src="https://github.com/user-attachments/assets/2ca64985-0e4d-4d2d-9e13-69c59a45faf6" width="600" />

I know literally nothing about Lua or contributing to open source, so chances I've made a mistake somewhere are not zero, please keep in mind.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/13994)
<!-- Reviewable:end -->

Commit:
black_hex & white_hex added

Commit:
minor fixes

Commit:
a few more black_hexes and white_hexes

Commit:
notification message

Commit:
0xcolor -> 0xCOLOR

Commit:
white_threshold

kerivin committed:
Merge branch 'koreader:master' into master

Commit:
more precise white threshold option setting

Commit:
bump base

Commit:
bump base

Commit:
bump base

Commit:
bump base

Commit:
cleanup after experiments

Commit:
cleanup after experiments

Commit:
bump base

kerivin committed:
Merge branch 'koreader:master' into master

Commit:
bump base

Commit:
Merge branch 'master' of https://github.com/kerivin/koreader

kerivin committed:
Merge branch 'master' of github.com:koreader/koreader

kerivin commented:
@Frenzie I was passing by again and am still not sure what's wrong with the tests. Maybe someone could take a look? Now I don't really know what to fix

Frenzie commented:
```

Frenzie commented:
This seems to indicate something shifted in ways it shouldn't. Sorry, I don't have time to investigate right now.

```
[ RUN      ] spec/front/unit/readerhighlight_spec.lua:146: Readerhighlight module highlight for PDF documents in page mode for scanned page with text layer should highlight word
spec/front/unit/readerhighlight_spec.lua:150: Expected objects to be equal.
Passed in:
(string) 'and'
Expected:
(string) 'Penn'
```

```
 documents in page mode for scanned page with text layer should highlight text
spec/front/unit/readerhighlight_spec.lua:157: Expected objects to be equal.
Passed in:
(string) 'upon missions into foreign countries, and accordingly they embarked for Hol land,'
Expected:
(string) 'to take care of the London'
```

hugleo commented:
Careful in koptinterface.lua:

I suggest placing the white_threshold parameter after the hinting parameter in the function:
 
```KoptInterface:renderPage(doc, pageno, rect, zoom, rotation, gamma, white_threshold, hinting)```

For example here white_threshold will be used instead of hinting:


https://github.com/koreader/koreader/blob/master/frontend/document/koptinterface.lua#L399C44-L399C86 

```return Document.renderPage(doc, pageno, rect, zoom, rotation, gamma, hinting)```

kerivin commented:
@hugleo It seems a good idea to keep `gamma` and `white_threshold` close since they have similar purpose unlike `hinting`. I think I changed all the places where `KoptInterface:renderPage` (and others with `white_threshold`) gets invoked, including `PdfDocument` where tests fail:
```
function PdfDocument:renderPage(pageno, rect, zoom, rotation, gamma, white_threshold, hinting)
    return self.koptinterface:renderPage(self, pageno, rect, zoom, rotation, gamma, white_threshold, hinting)
end
```
If I used wrong argument somewhere, white threshold wouldn`t work at all I guess, but it works locally, both with MuPDF and libk2pdfopt.
Logic I followed (if I understand Lua correctly, feel free to object):

`PdfDocument:renderPage` gets invoked from this:
```
function Document:drawPage(target, x, y, rect, pageno, zoom, rotation, gamma, white_threshold)
    local tile = self:renderPage(pageno, rect, zoom, rotation, gamma, white_threshold)
    ...
end
```

`Document:drawPage` from this:
```
function KoptInterface:drawPage(doc, target, x, y, rect, pageno, zoom, rotation, gamma, white_threshold)
    ...
    Document.drawPage(doc, target, x, y, rect, pageno, zoom, rotation, gamma, white_threshold)
    ...
end
```

`KoptInterface:drawPage` from this (still going by PDF call stack for the sake of an example):
```
function PdfDocument:drawPage(target, x, y, rect, pageno, zoom, rotation, gamma, white_threshold)
    return self.koptinterface:drawPage(self, target, x, y, rect, pageno, zoom, rotation, gamma, white_threshold)
end
```

And finally `PdfDocument:drawPage` invoked from here:
```
function ReaderView:drawSinglePage(bb, x, y)
    self.document:drawPage(
        bb,
        x + self.state.offset.x,
        y + self.state.offset.y,
        self.visible_area,
        self.state.page,
        self.state.zoom,
        self.state.rotation,
        self.state.gamma,
        self.state.white_threshold)
    UIManager:nextTick(self.emitHintPageEvent)
end
```

Obviously it's not the only call stack possible, but this is what I guess how the renderer usually works. I don't see where I could have accidentally skipped adding `white_threshold` argument, but since tests fail I guess I have...

My guesses:
1. I could have broken configuration reading or something? Since I've added the new UI parameter not at the end of the configuration list but after `gamma`.
2. Forgotten `white_threshold` somewhere
3. Wrong argument order

hugleo commented:
Could it be here for example?
https://github.com/kerivin/koreader/blob/master/frontend/document/koptinterface.lua#L389C24-L389C69
from ```local tile = Document.renderPage(doc, 1, nil, zoom, 0, 1.0, 255)```
to? ```local tile = Document.renderPage(doc, 1, nil, zoom, 0, 255, 1.0)```

kerivin committed:
kc:setWhite(doc.configurable.white_threshold) fix

kerivin committed:
separate setting paint_white for libk2pdfopt

kerivin committed:
libk2pdfopt commit hash update

kerivin committed:
update base

kerivin committed:
update base

spinktvis commented:
Question, with this added functionality would it then also be possible to entirely disable or discard rendering the "grey page background" layer in the PDF? And in turn make KOReader render a PDF faster and for example with Reflow enabled closer to for Moon Reader or Librera PDF to epub functionality?
